main/reference/vstorage-ref.md

```diff
@@ -1,13 +1,83 @@
-# VStorage Reference
+# VStorage in Agoric
 
-See also:
+In the Agoric platform, VStorage, or a Virtual Storage, is a key-value store that:
 
-- [Querying VStorage](/guides/getting-started/contract-rpc#querying-vstorage)
-- [Publishing to chainStorage](../guides/zoe/pub-to-storage)
-- [x/vstorage module](https://github.com/Agoric/agoric-sdk/tree/003f0c2232815a8d64a3f9a5b05521a10160ce34/golang/cosmos/x/vstorage#readme)
+- provides a **read-only interface for clients of the consensus layer**. Clients can query the stored data using specific paths.
+- is organized in a **hierarchical, path-based structure**. Each fragment of data is stored at a specific path, and clients can query these paths to retrieve the data.
+- can be written through a specialized API called **[chainStorage](https://docs.agoric.com/guides/zoe/pub-to-storage.html#publishing-to-chainstorage)**. This API allows the VM to manage data storage in a controlled and secure manner.
+
+![vstorage query diagram](../guides/getting-started/assets/vstorage-brand-q.svg)
+
+## VStorage Hierarchy
+
+VStorage is structured as a tree with paths and nodes that store the actual data. This design facilitates easy querying while ensuring data consistency and security.
+
+For example, here is an example of [Inter Protocol key structure](https://github.com/Agoric/agoric-sdk/tree/agoric-upgrade-13/packages/inter-protocol#reading-data-off-chain) inside Agoric:
+
+```
+- published
+    - vaultFactory
+        - governance
+        - metrics
+        - managers
+            - manager0
+                - metrics
+                - governance
+                - vaults
+                    - vault0
+```
+
+You can then use this hierarchy to access the data fragments.
+```sh
+# lists vaults
+$ agd query vstorage keys 'published.vaultFactory.managers.manager0.vaults'
+children:
+- vault0
+```
+
+## VStorage with CLI
+VStorage can be accessed via CLI using:
+
+```sh
+$ agd /[--node {url}] query vstorage {path}
+```
+
+For example:
+```sh
+$ agd query vstorage path published.agoricNames
+children:
+- brand
+- installation
+- instance
+...
```

Comment:
Ah. yes. this is the sort of material that I'd like to see in https://docs.agoric.com/guides/agoric-cli/agd-query-tx.html

The `--node` flag is discussed more thoroughly there. I don't see any reference to it here, so you might as well drop it here.

What's the `/` in `/[--node {url}]` by the way?

Reply:
`/` was a typo. Fixing it in a commit. 

```
The query goes to a local node at tcp://localhost:26657 by default. To use another node:
$ agd status --node https://devnet.rpc.agoric.net:443
```

Can you confirm this is same for `query` as well? I am not sure about it. There is no mention about `--node` parameter there. Maybe I can add information there and then refer it to that.

Reply:
yes, `--node` applies to all `agd query` and `agd tx` sub-commands, IME.

The help text is oddly inconsistent, though: `agd query --help | grep node` is empty, as is `agd query vstorage --help | grep node` but:

```console
$ agd query tx --help | grep node
      --node string        <host>:<port> to Tendermint RPC interface for this chain (default "tcp://localhost:26657")
```

@gibson042 Do you know of some reason why the help text is inconsistent? If not, I suppose we owe an issue?

Reply:
Yes, an issue would be good. I'm pretty sure this is because the [Cobra](https://cobra.dev/) library responsible for CLI option definition shows help for applicable options defined higher in the command hierarchy (so-called ["persistent flags"](https://cobra.dev/#persistent-flags)) only at an invocable command (which `agd query tx` is but e.g. `agd query vstorage` is not because it must be followed by "children" or "data" or "path" and `agd query params` is not because it must be followed by "subspace").

But there should be a way to do it, because "--chain-id" seems to work that way (e.g., it shows in both e.g. `agd tx --help` and [as a Global Flag] in `agd tx gov --help`).